File: src/commands/skills.ts
import { existsSync, mkdirSync, writeFileSync } from "fs";
import { join } from "path";
import type { Command } from "commander";
import { skillContent, commandsReference } from "../skills/content.ts";

type AgentType = "claude" | "copilot" | "codex";

const AGENT_SKILL_DIRS: Record<AgentType, string> = {
  claude: ".claude/skills/jenkins",
  copilot: ".github/skills/jenkins",
  codex: ".codex/skills/jenkins",
};

const VALID_AGENTS = Object.keys(AGENT_SKILL_DIRS) as AgentType[];

interface InstallResult {
  agent: AgentType;
  path: string;
  action: "installed" | "skipped" | "updated";
}

export function registerSkillCommands(program: Command): void {
  program
    .command("install-skill")
    .description("Install AI agent skill for the Jenkins CLI into the current directory")
    .option("--force", "overwrite existing skill files", false)
    .option("--agent <type>", "target a specific agent (claude, copilot, codex) instead of auto-detecting")
    .action(async (opts: { force: boolean; agent?: string }) => {
      const cwd = process.cwd();

      if (opts.agent && !VALID_AGENTS.includes(opts.agent as AgentType)) {
        throw new Error(`Unknown agent type: ${opts.agent}. Use ${VALID_AGENTS.join(", ")}.`);
      }

      const agents: AgentType[] = opts.agent
        ? [opts.agent as AgentType]
        : detectAgents(cwd);

      if (agents.length === 0) {
        const dirs = VALID_AGENTS.map((a) => agentRootDir(a) + "/").join(", ");
        throw new Error(
          `No AI agent directories detected. Checked for: ${dirs}\n` +
          `Use --agent <${VALID_AGENTS.join("|")}> to install for a specific agent.`,
        );
      }

      const results = agents.map((agent) => installSkill(cwd, agent, opts.force));

      for (const r of results) {
        const label = r.agent.charAt(0).toUpperCase() + r.agent.slice(1);
        switch (r.action) {
          case "installed":
            console.log(`${label}: installed → ${r.path}`);
            break;
          case "updated":
            console.log(`${label}: updated → ${r.path}`);
            break;
          case "skipped":
            console.log(`${label}: already exists → ${r.path} (use --force to overwrite)`);
            break;
        }
      }
    });
}

function detectAgents(cwd: string): AgentType[] {
  return VALID_AGENTS.filter((agent) =>
    existsSync(join(cwd, agentRootDir(agent))),
  );
}

function agentRootDir(agent: AgentType): string {
  return AGENT_SKILL_DIRS[agent].split("/")[0]!;
}

function installSkill(cwd: string, agent: AgentType, force: boolean): InstallResult {
  const skillDir = join(cwd, AGENT_SKILL_DIRS[agent]);
  const skillPath = join(skillDir, "SKILL.md");
  const refsDir = join(skillDir, "references");
  const refsPath = join(refsDir, "commands.md");
  const existed = existsSync(skillPath);

  if (existed && !force) {
    return { agent, path: skillPath, action: "skipped" };
  }

  mkdirSync(refsDir, { recursive: true });
  writeFileSync(skillPath, skillContent(AGENT_SKILL_DIRS[agent]));
  writeFileSync(refsPath, commandsReference());

  return { agent, path: skillPath, action: existed ? "updated" : "installed" };
}

File: cmd/skills.go
package cmd

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/spf13/cobra"

	"github.com/nwp/jenkins-cli/internal/skills"
)

type agentConfig struct {
	name string
	dir  string
}

var supportedAgents = []agentConfig{
	{name: "claude", dir: ".claude"},
	{name: "copilot", dir: ".github"},
	{name: "codex", dir: ".codex"},
}

func newInstallSkillCmd() *cobra.Command {
	var force bool
	var agentType string

	c := &cobra.Command{
		Use:   "install-skill",
		Short: "Install the Jenkins CLI AI agent skill into the current directory",
		RunE: func(cmd *cobra.Command, args []string) error {
			cwd, err := os.Getwd()
			if err != nil {
				return fmt.Errorf("get working directory: %w", err)
			}

			var targets []agentConfig
			if agentType != "" {
				for _, a := range supportedAgents {
					if a.name == agentType {
						targets = []agentConfig{a}
						break
					}
				}
				if len(targets) == 0 {
					return fmt.Errorf("unknown agent type %q (supported: claude, copilot, codex)", agentType)
				}
			} else {
				// Auto-detect
				for _, a := range supportedAgents {
					if _, err := os.Stat(filepath.Join(cwd, a.dir)); err == nil {
						targets = append(targets, a)
					}
				}
				if len(targets) == 0 {
					return fmt.Errorf("no supported AI agent directories found (.claude/, .github/, .codex/); use --agent to specify one")
				}
			}

			for _, a := range targets {
				status, err := installSkillFor(cwd, a, force)
				if err != nil {
					fmt.Fprintf(os.Stderr, "ERROR: install for %s: %v\n", a.name, err)
					continue
				}
				fmt.Printf("[%s] %s\n", a.name, status)
			}
			return nil
		},
	}

	c.Flags().BoolVar(&force, "force", false, "overwrite existing skill files")
	c.Flags().StringVar(&agentType, "agent", "", "target agent (claude, copilot, codex)")
	return c
}

func installSkillFor(cwd string, a agentConfig, force bool) (string, error) {
	skillDir := filepath.Join(cwd, a.dir, "skills", "jenkins")
	refDir := filepath.Join(skillDir, "references")

	if err := os.MkdirAll(refDir, 0o755); err != nil {
		return "", fmt.Errorf("create skill directory: %w", err)
	}

	skillFile := filepath.Join(skillDir, "SKILL.md")
	refFile := filepath.Join(refDir, "commands.md")

	skillStatus := "installed"
	if _, err := os.Stat(skillFile); err == nil {
		if !force {
			skillStatus = "skipped (already exists)"
		} else {
			skillStatus = "updated"
		}
	}

	if skillStatus != "skipped (already exists)" {
		if err := os.WriteFile(skillFile, []byte(skills.SkillContent(skillDir)), 0o644); err != nil {
			return "", fmt.Errorf("write SKILL.md: %w", err)
		}
		if err := os.WriteFile(refFile, []byte(skills.CommandsReference()), 0o644); err != nil {
			return "", fmt.Errorf("write commands.md: %w", err)
		}
	}

	return skillStatus, nil
}
